io/src/auction.rs

```diff
@@ -0,0 +1,220 @@
+use crate::io::*;
+use core::cmp::min;
+use gstd::{exec, msg, prelude::*, ActorId};
+use nft_io::{NFTAction, NFTEvent};
+use primitive_types::U256;
+
+#[derive(Debug, Decode, Default, Encode, TypeInfo, Clone)]
+#[codec(crate = gstd::codec)]
+#[scale_info(crate = gstd::scale_info)]
+pub enum Status {
+    #[default]
+    None,
+    IsRunning,
+    Purchased {
+        price: u128,
+    },
+    Expired,
+    Stopped,
+}
+
+#[derive(Debug, Clone, Default, Encode, Decode, TypeInfo)]
+pub struct NFT {
+    pub token_id: U256,
+    pub owner: ActorId,
+    pub contract_id: ActorId,
+}
+
+#[derive(Debug, Clone, Default, Encode, Decode, TypeInfo)]
+#[codec(crate = gstd::codec)]
+#[scale_info(crate = gstd::scale_info)]
+pub struct Auction {
+    pub owner: ActorId,
+    pub nft: NFT,
+    pub starting_price: u128,
+    pub discount_rate: u128,
+    pub status: Status,
+    pub started_at: u64,
+    pub expires_at: u64,
+}
+
+impl Auction {
```

Comment:
Usually, we hadn't `impl`s in `io` crate. Isn't it better to keep business logic in one crate?